src/Chapter3.hs

```diff
@@ -480,6 +525,80 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
 
+data Building
+    = Church
+    | Library
+    deriving (Show)
+
+data Castle
+    = Nope
+    | CastleWithWalls String
+    | Castle String
+    deriving (Show)
+
+
+data House
+    = One
+    | Two
+    | Three
+    | Four
+    deriving (Show)
+
+
+data City = City
+    {cityBuilding :: Building
+    ,cityCastle   :: Castle
+    ,cityHouses   :: [House]
+    }
+    deriving (Show)
+
+
+-- buildCastle :: Castle -> City -> City
+-- buildCastle newCastle city = city {cityCastle = newCastle}
+
+{- | buildCastle — build a castle in the city. If the city already has a castle,
+   the old castle is destroyed, and the new castle with the __new name__ is built
+-}
+buildCastle :: String -> City -> City
+buildCastle name city@(City _ castle _) = case castle of
+    CastleWithWalls _ -> city {cityCastle = CastleWithWalls name}
+    Castle _          -> city {cityCastle = Castle name}
+    _                 -> city
```

Comment:
In this case you also need to build a castle, because the city doesn't have a `Castle` here

Reply:
Yes, I don't know why I make a special case here.